### PR TITLE
feat: Add Mistral Amazon Bedrock support 

### PR DESCRIPTION
**Pull request by vblagoje**
### Why:
Adds Mistral models in the Amazon Bedrock integration within the haystack-core-integration ecosystem. This addition allows the use of Mistral models for generating responses and enables handling chat message for applications requiring conversational agents.

### What:
- **Introduced MistralAdapter**: A new adapter for handling requests and responses for Mistral models.
- **Introduced MistralChatAdapter**: Extends support for chat interfaces utilizing Mistral models, including handling of chat templates and generation parameters specific to Mistral.
- **Integration with Amazon Bedrock Chat Generator**: Registered MistralChatAdapter to enable its selection based on model patterns within the Amazon Bedrock Chat Generator mechanism.
- **Test Coverage**: Added comprehensive testing for both MistralAdapter and MistralChatAdapter to validate functionality and response handling.

### How can it be used:
- **AmazonBedrockGenerator **:
```python
from haystack_integrations.components.generators.amazon_bedrock import AmazonBedrockGenerator

abg = AmazonBedrockGenerator("mistral.mixtral-8x7b-instruct-v0:1", max_length=1024)
result = abg.run(prompt="Tell me a bit about Berlin")
print(result)
  ```
- **AmazonBedrockChatGenerator**:
```python
from haystack.components.generators.utils import print_streaming_chunk
from haystack.dataclasses import ChatMessage

from haystack_integrations.components.generators.amazon_bedrock import  AmazonBedrockChatGenerator
messages = [ChatMessage.from_system("You are a helpful, respectful and honest assistant, answer in German only"),
            ChatMessage.from_user("What's Natural Language Processing?")]
abg = AmazonBedrockChatGenerator("mistral.mixtral-8x7b-instruct-v0:1", streaming_callback=print_streaming_chunk)
result = abg.run(messages=messages)
print(result)
  ```

### How did you test it:
- Executed unit tests for both adapters, ensuring correct preparation of request bodies and accurate parsing of model responses. 
- Conducted integration tests simulating real chat flows, verifying that responses from the Mistral models align with expectations and maintain conversational context.
- Verified the handling of custom generation parameters and their precedence over default parameters.

### Notes for the reviewer:
- Pay special attention to how the new adapters integrate with the existing framework, especially regarding the pattern matching used to select these adapters based on the model name.
- Review the template handling in MistralChatAdapter, ensuring compatibility with system messages and compliance with expected Mistral model inputs.
- For further enhancement, consider the possibility of adapting or extending the test scenarios to cover more diversified chat flows and generation parameters.

### CODE REVIEW

**Comment by vblagoje** [comment]
@masci don't start review yet, I'll do another round of checks tomorrow

**Comment by vblagoje** [comment]
Yes saw that announcement as well - will add and test it today and push for integration of this PR. Thanks for the ping @lambda-science 🙏

**Comment by lambda-science** [comment]
Following this new integration with interest !
Btw yesterday Mistral large has been anounced and added to Bedrock: https://aws.amazon.com/blogs/aws/tackle-complex-reasoning-tasks-with-mistral-large-now-available-on-amazon-bedrock/ -> https://github.com/deepset-ai/haystack-core-integrations/pull/632/files#diff-0d179160451053ac5bbdb4097672903d96568752612edd2510f5c432bb994e5bR17

**Comment by lambda-science** [comment]
> Yes saw that announcement as well - will add and test it today and push for integration of this PR. Thanks for the ping @lambda-science 🙏

I got confused with interface, I didn't meant to start a review and deleted my comment so I put it back here 😂

"Interested with this new integration and Bedrock anounced yesterday that the new Mistral Large model is avaliable (https://aws.amazon.com/blogs/aws/tackle-complex-reasoning-tasks-with-mistral-large-now-available-on-amazon-bedrock/) 
Relevant line of code: https://github.com/deepset-ai/haystack-core-integrations/pull/632/files#diff-0d179160451053ac5bbdb4097672903d96568752612edd2510f5c432bb994e5bR17 "

**Comment by vblagoje** [comment]
@lambda-science the upcoming release of amazon bedrock supports all models. See [c60e4ea](https://github.com/deepset-ai/haystack-core-integrations/pull/632/commits/c60e4ea4a2c7b359fde1e380681249549f928e2c) for models we test

**Comment by vblagoje** [comment]
@masci would you please have a look at this PR now?

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/adapters.py [hunk 1]
```diff
@@ -271,6 +271,158 @@ def _to_anthropic_message(self, m: ChatMessage) -> Dict[str, Any]:
         return {"content": [{"type": "text", "text": m.content}], "role": m.role.value}
 
 
+class MistralChatAdapter(BedrockModelChatAdapter):
+    """
+    Model adapter for the Mistral chat model.
+    """
+
+    chat_template = (
```
Comment (masci) [inline integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/adapters.py:279]: Can we add tests for the logic inside this template? Also we should probably move to triple quotes to ease copy/paste of the template code and readability

**Review by masci — CHANGES_REQUESTED** [review]

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/adapters.py (hunk 1 above)
Comment (vblagoje) [reply integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/adapters.py:279]: Sure, good point on both!

**Comment by vblagoje** [comment]
@masci these templates are sensitive to whitespace so I can either oneline them, revert back as they were or use some utils liket his [one](https://stackoverflow.com/a/48112903)

**Comment by masci** [comment]
Not sure I get the problem, triple quotes should preserve inner whitespaces, and we can `strip` the template if the leading and trailing spaces are a problem, like:
```py
    template = """
This
  will
be
      preserved
""".strip()  # <- this will ensure leading and trailing \n are removed

```

**Comment by vblagoje** [comment]
> Not sure I get the problem, triple quotes should preserve inner whitespaces, and we can `strip` the template if the leading and trailing spaces are a problem, like:
> 
> ```python
>     template = """
> This
>   will
> be
>       preserved
> """.strip()  # <- this will ensure leading and trailing \n are removed
> ```

Apologies for not being clear, we need to do the opposite, remove all leading/ending whitespaces between lines and \n\r as well. See last few commits. This way we get to preserve the visual cues of jinja template structure and also being able to use it without these whitespace hiccups. The inspiration for this template comes from this [one](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1/blob/main/tokenizer_config.json) however, it doesn't support system messages - that's the part we changed. Note how hf also has it as one-liner for the same reasons I believe.

**Comment by vblagoje** [comment]
ping @masci

**Pull request merged**